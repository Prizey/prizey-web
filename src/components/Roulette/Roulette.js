import React, { useState } from 'react'
import { withStyles } from '@material-ui/core/styles'
import useInterval from 'hooks/useInterval'
import posed from 'react-pose'

const styles = () => ({
  img: {
    height: '300px',
    position: 'absolute',
    width: '300px',
  },
  root: {
    height: '300px',
    width: '300px',
  },
})

const PosedImg = posed.img({
  enter: {
    opacity: 1,
    transition: {
      x: { duration: 1 },
    },
    x: 0,
  },
  leave: {
    opacity: 0,
    transition: {
      x: { duration: 1 },
    },
    x: -9999,
  },
})

export default withStyles(styles)(({ classes, speed, onSelectItem, data }) => {
  const [next, setNext] = useState(0)

  useInterval(() => {
    setNext((next + 1) % data.length)
  }, speed)

  return (
    <div className={classes.root}>
      {data.map((item, id) => (
        <PosedImg
          key={id}
          onClick={() => onSelectItem(item)}
          pose={next === id ? 'enter' : 'leave'}
          src={item}
          alt="unsplash"
          className={classes.img}
        />
      ))}
    </div>
  )
})
