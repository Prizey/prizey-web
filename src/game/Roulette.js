import React, { useState, useLayoutEffect } from 'react'
import { withStyles } from '@material-ui/core/styles'
import posed from 'react-pose'

export const getItemSpeed = (price, speed, multiplier) => {
  const PRICE_FACTOR = process.env.REACT_APP_PRICE_FACTOR || 1
  const deltaSpeed = (PRICE_FACTOR / price) * multiplier * speed
  return Math.max(deltaSpeed, 80)
}

const styles = () => ({
  img: {
    maxWidth: '300px',
    position: 'absolute',
    width: '100%',
  },
  root: {
    height: '300px',
    maxWidth: '300px',
    position: 'relative',
    width: '100%',
  },
})

export const enterTransition = ({ speed }) => ({
  opacity: {
    duration: Math.min(speed - 50, 500),
    loop: 100,
  },
  x: { duration: 10 },
})

const PosedImg = posed.img({
  enter: {
    opacity: 1,
    transition: enterTransition,
    x: 0,
  },
  leave: {
    opacity: 0,
    transition: {
      opacity: { duration: 10, ease: 'easeOut' },
      x: { duration: 10 },
    },
    x: -9999,
  },
})

export default withStyles(styles)(({ classes, speed, onSelectItem, data }) => {
  const [next, setNext] = useState(0)
  const multiplier = 2

  useLayoutEffect(() => {
    const item = data[next]
    const timeout = setTimeout(() => {
      setNext((next + 1) % data.length)
    }, getItemSpeed(item.price, speed, multiplier))

    return () => clearTimeout(timeout)
  }, [speed, data, next, setNext])

  return (
    <div className={classes.root}>
      {data.map((item, id) => (
        <div
          key={`${item.id}-${id}`}
          testID={`rouletteItem-${id}`}
          onMouseDown={() => onSelectItem(item)}
        >
          <PosedImg
            pose={next === id ? 'enter' : 'leave'}
            src={item.image}
            alt={item.title}
            className={classes.img}
            speed={getItemSpeed(item.price, speed, multiplier)}
          />
        </div>
      ))}
    </div>
  )
})
