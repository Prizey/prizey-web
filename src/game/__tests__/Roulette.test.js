import React from 'react'
import renderer, { act } from 'react-test-renderer'
import Roulette, { getItemSpeed, enterTransition } from '../Roulette'

jest.useFakeTimers()

const mockList = [
  { id: 1, image: '/mocks/trump-mask.png', price: 10, title: 'trump' },
  { id: 2, image: '/mocks/sweatshirt.png', price: 10, title: 'sweatshirt' },
  { id: 3, image: '/mocks/shoe.png', price: 10, title: 'shoe' },
]

it('gets the correct item speed', () => {
  const speed = getItemSpeed(10, 1000, 1)
  expect(speed).toEqual(1000)
})

it('gets the correct transition information', () => {
  const transition = enterTransition({ speed: 250 })
  expect(transition).toEqual({
    opacity: { duration: 200 },
    x: { duration: 10 },
  })
})

it('renders correctly', () => {
  const tree = renderer
    .create(<Roulette speed={2000} data={mockList} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('the tap works', () => {
  const onTap = jest.fn()
  const tree = renderer.create(
    <Roulette speed={250} data={mockList} onSelectItem={onTap} />,
  ).root

  tree.findByProps({ 'aria-label': 'rouletteItem-0' }).props.onMouseDown()

  expect(onTap).toHaveBeenCalledWith(mockList[0])
})

it('go to the next image', () => {
  const tree = renderer.create(<Roulette speed={10000} data={mockList} />).root

  act(() => {
    jest.runOnlyPendingTimers()
  })
  expect(tree.findByProps({ pose: 'enter' }).props.src).toEqual(
    mockList[1].image,
  )
})
