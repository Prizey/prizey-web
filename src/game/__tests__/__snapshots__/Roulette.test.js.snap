// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="Component-root-2"
>
  <div
    aria-label="rouletteItem-0"
    onMouseDown={[Function]}
  >
    <img
      alt="trump"
      className="Component-img-1"
      pose="enter"
      speed={5000}
      src="/mocks/trump-mask.png"
    />
  </div>
  <div
    aria-label="rouletteItem-1"
    onMouseDown={[Function]}
  >
    <img
      alt="sweatshirt"
      className="Component-img-1"
      pose="leave"
      speed={5000}
      src="/mocks/sweatshirt.png"
    />
  </div>
  <div
    aria-label="rouletteItem-2"
    onMouseDown={[Function]}
  >
    <img
      alt="shoe"
      className="Component-img-1"
      pose="leave"
      speed={5000}
      src="/mocks/shoe.png"
    />
  </div>
</div>
`;
